core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ProviderContractNegotiationManagerImpl.java

```diff
@@ -292,20 +295,19 @@ private int checkProviderOffering() {
                     .build();
 
             //TODO protocol-independent response type?
-            var response = dispatcherRegistry.send(Object.class, contractOfferRequest, () -> null);
-
-            if (response.isCompletedExceptionally()) {
-                negotiation.transitionOffering();
-                negotiationStore.save(negotiation);
-                monitor.debug(format("[Provider] Failed to send contract offer with id %s. ContractNegotiation %s stays in state %s.",
-                        currentOffer.getId(), negotiation.getId(), ContractNegotiationStates.from(negotiation.getState())));
-                continue;
-            }
-
-            negotiation.transitionOffered();
-            negotiationStore.save(negotiation);
-            monitor.debug(String.format("[Provider] ContractNegotiation %s is now in state %s.",
-                    negotiation.getId(), ContractNegotiationStates.from(negotiation.getState())));
+            dispatcherRegistry.send(Object.class, contractOfferRequest, () -> null)
+                    .whenComplete((response, throwable) -> {
+                        if (throwable == null) {
+                            negotiation.transitionOffered();
+                            negotiationStore.save(negotiation);
+                            monitor.debug(String.format("[Provider] ContractNegotiation %s is now in state %s.",
+                                    negotiation.getId(), ContractNegotiationStates.from(negotiation.getState())));
+                        } else {
+                            String message = format("[Provider] Failed to send contract offer with id %s. ContractNegotiation %s stays in state %s.",
```

Comment:
See comments on `ConsumerContractnegotiationManagerImpl`. For all changes in this class.